// openapi2aspida/dist/buildV3.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var converters_1 = require("./builderUtils/converters");
var props2String_1 = require("./builderUtils/props2String");
var resolvers_1 = require("./builderUtils/resolvers");
var getDirName_1 = __importDefault(require("./builderUtils/getDirName"));
var schemas2Props_1 = __importDefault(require("./builderUtils/schemas2Props"));
var parameters2Props_1 = __importDefault(require("./builderUtils/parameters2Props"));
var requestBodies2Props_1 = __importDefault(require("./builderUtils/requestBodies2Props"));
var responses2Props_1 = __importDefault(require("./builderUtils/responses2Props"));
var methodNames = ['get', 'post', 'put', 'delete', 'head', 'options', 'patch'];
var getParamsList = function (openapi, params) { return (params === null || params === void 0 ? void 0 : params.map(function (p) { return ((0, converters_1.isRefObject)(p) ? (0, resolvers_1.resolveParamsRef)(openapi, p.$ref) : p); })) || []; };
exports.default = (function (openapi) {
    var _a, _b, _c, _d, _e, _f;
    var files = [];
    var schemas = (0, schemas2Props_1.default)((_a = openapi.components) === null || _a === void 0 ? void 0 : _a.schemas, openapi) || [];
    var parameters = (0, parameters2Props_1.default)((_b = openapi.components) === null || _b === void 0 ? void 0 : _b.parameters, openapi, false) || [];
    var requestBodies = (0, requestBodies2Props_1.default)((_c = openapi.components) === null || _c === void 0 ? void 0 : _c.requestBodies) || [];
    var responses = (0, responses2Props_1.default)((_d = openapi.components) === null || _d === void 0 ? void 0 : _d.responses) || [];
    files.push.apply(files, __spreadArray([], __read(Object.keys(openapi.paths)
        .map(function (path) {
        var methodProps = Object.keys(openapi.paths[path]).filter(function (method) {
            return methodNames.includes(method);
        });
        var file = __spreadArray(__spreadArray([], __read(path
            .replace(/\/$/, '')
            .split('/')
            .slice(1)
            .map(function (p) {
            return (0, getDirName_1.default)(p, __spreadArray(__spreadArray([], __read(getParamsList(openapi, openapi.paths[path].parameters)), false), __read(methodProps.reduce(function (prev, c) {
                var _a;
                return __spreadArray(__spreadArray([], __read(prev), false), __read(getParamsList(openapi, (_a = openapi.paths[path][c]) === null || _a === void 0 ? void 0 : _a.parameters)), false);
            }, [])), false), false);
        })), false), [
            'index'
        ], false);
        var methods = methodProps
            .map(function (method) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
            var target = openapi.paths[path][method];
            if (target.deprecated)
                return null;
            var params = [];
            if (target.parameters || openapi.paths[path].parameters) {
                var reqRefHeaders_1 = [];
                var reqHeaders_1 = [];
                var refQuery_1 = [];
                var query_1 = [];
                var queryRequired_1 = false;
                __spreadArray(__spreadArray([], __read((openapi.paths[path].parameters || [])), false), __read((target.parameters || [])), false).forEach(function (p) {
                    var _a, _b, _c, _d, _e;
                    if ((0, converters_1.isRefObject)(p)) {
                        var ref = (0, resolvers_1.resolveParamsRef)(openapi, p.$ref);
                        var val = {
                            isArray: false,
                            isEnum: false,
                            nullable: false,
                            description: (_a = ref.description) !== null && _a !== void 0 ? _a : null,
                            value: (0, converters_1.$ref2Type)(p.$ref)
                        };
                        switch (ref.in) {
                            case 'header':
                                reqRefHeaders_1.push(val);
                                break;
                            case 'query':
                                refQuery_1.push(val);
                                queryRequired_1 = queryRequired_1 || ((_b = ref.required) !== null && _b !== void 0 ? _b : false);
                                break;
                            default:
                                break;
                        }
                    }
                    else {
                        var value = (0, converters_1.schema2value)(p.schema, false);
                        if (!value)
                            return;
                        var prop = {
                            name: (0, converters_1.getPropertyName)(p.name),
                            required: (_c = p.required) !== null && _c !== void 0 ? _c : false,
                            description: (_d = p.description) !== null && _d !== void 0 ? _d : null,
                            values: [value]
                        };
                        switch (p.in) {
                            case 'header':
                                reqHeaders_1.push(prop);
                                break;
                            case 'query':
                                query_1.push(prop);
                                queryRequired_1 = queryRequired_1 || ((_e = p.required) !== null && _e !== void 0 ? _e : false);
                                break;
                            default:
                                break;
                        }
                    }
                });
                if (reqHeaders_1.length || reqRefHeaders_1.length) {
                    params.push({
                        name: 'reqHeaders',
                        required: false,
                        description: null,
                        values: __spreadArray(__spreadArray([], __read(reqRefHeaders_1), false), __read((reqHeaders_1.length
                            ? [
                                {
                                    isArray: false,
                                    isEnum: false,
                                    nullable: false,
                                    description: null,
                                    value: reqHeaders_1
                                }
                            ]
                            : [])), false)
                    });
                }
                if (refQuery_1.length || query_1.length) {
                    params.push({
                        name: 'query',
                        required: queryRequired_1,
                        description: null,
                        values: __spreadArray(__spreadArray([], __read(refQuery_1), false), __read((query_1.length
                            ? [
                                {
                                    isArray: false,
                                    isEnum: false,
                                    nullable: false,
                                    description: null,
                                    value: query_1
                                }
                            ]
                            : [])), false)
                    });
                }
            }
            if (target.responses) {
                var code = Object.keys(target.responses).find(function (code) { return code.match(/^(20\d|30\d)$/); });
                if (code) {
                    params.push({
                        name: 'status',
                        required: true,
                        description: null,
                        values: [
                            {
                                isArray: false,
                                isEnum: false,
                                nullable: false,
                                description: null,
                                value: code
                            }
                        ]
                    });
                    var res = target.responses[code];
                    var ref_1 = (0, converters_1.isRefObject)(res) ? (0, resolvers_1.resolveResRef)(openapi, res.$ref) : res;
                    var content = (_b = (ref_1.content &&
                        ((_a = Object.entries(ref_1.content).find(function (_a) {
                            var _b = __read(_a, 1), key = _b[0];
                            return key.startsWith('application/');
                        })) === null || _a === void 0 ? void 0 : _a[1]))) !== null && _b !== void 0 ? _b : (_c = ref_1.content) === null || _c === void 0 ? void 0 : _c[Object.keys(ref_1.content)[0]];
                    if (content === null || content === void 0 ? void 0 : content.schema) {
                        var val = (0, converters_1.schema2value)(content.schema, true, true);
                        val &&
                            params.push({
                                name: 'resBody',
                                required: true,
                                description: ref_1.description,
                                values: [val]
                            });
                    }
                    if (ref_1.headers) {
                        params.push({
                            name: 'resHeaders',
                            required: true,
                            description: null,
                            values: [
                                {
                                    isArray: false,
                                    isEnum: false,
                                    nullable: false,
                                    description: null,
                                    value: Object.keys(ref_1.headers)
                                        .map(function (header) {
                                        var _a;
                                        var headerData = ref_1.headers[header];
                                        var val = (0, converters_1.isRefObject)(headerData)
                                            ? {
                                                isArray: false,
                                                isEnum: false,
                                                description: null,
                                                value: (0, converters_1.$ref2Type)(headerData.$ref)
                                            }
                                            : (0, converters_1.schema2value)(headerData.schema, true);
                                        return (val && {
                                            name: (0, converters_1.getPropertyName)(header),
                                            required: (0, converters_1.isRefObject)(headerData)
                                                ? true
                                                : (_a = headerData.required) !== null && _a !== void 0 ? _a : true,
                                            description: (0, converters_1.isRefObject)(headerData)
                                                ? null
                                                : headerData.description,
                                            values: [val]
                                        });
                                    })
                                        .filter(function (v) { return !!v; })
                                }
                            ]
                        });
                    }
                }
            }
            if (target.requestBody) {
                var reqFormat = '';
                var reqBody = null;
                var required = true;
                var description = null;
                if ((0, converters_1.isRefObject)(target.requestBody)) {
                    var ref = (0, resolvers_1.resolveReqRef)(openapi, target.requestBody.$ref);
                    if ((_d = ref.content['multipart/form-data']) === null || _d === void 0 ? void 0 : _d.schema) {
                        reqFormat = 'FormData';
                    }
                    else if ((_e = ref.content['application/x-www-form-urlencoded']) === null || _e === void 0 ? void 0 : _e.schema) {
                        reqFormat = 'URLSearchParams';
                    }
                    reqBody = {
                        isArray: false,
                        isEnum: false,
                        nullable: false,
                        description: null,
                        value: (0, converters_1.$ref2Type)(target.requestBody.$ref)
                    };
                    required = (_f = ref.required) !== null && _f !== void 0 ? _f : true;
                    description = (_g = ref.description) !== null && _g !== void 0 ? _g : null;
                }
                else {
                    required = (_h = target.requestBody.required) !== null && _h !== void 0 ? _h : true;
                    description = (_j = target.requestBody.description) !== null && _j !== void 0 ? _j : null;
                    if ((_k = target.requestBody.content['multipart/form-data']) === null || _k === void 0 ? void 0 : _k.schema) {
                        reqFormat = 'FormData';
                        reqBody = (0, converters_1.schema2value)(target.requestBody.content['multipart/form-data'].schema, true);
                    }
                    else if ((_l = target.requestBody.content['application/x-www-form-urlencoded']) === null || _l === void 0 ? void 0 : _l.schema) {
                        reqFormat = 'URLSearchParams';
                        reqBody = (0, converters_1.schema2value)(target.requestBody.content['application/x-www-form-urlencoded'].schema, true);
                    }
                    else {
                        var content = target.requestBody.content &&
                            ((_m = Object.entries(target.requestBody.content).find(function (_a) {
                                var _b = __read(_a, 1), key = _b[0];
                                return key.startsWith('application/');
                            })) === null || _m === void 0 ? void 0 : _m[1]);
                        if (content === null || content === void 0 ? void 0 : content.schema)
                            reqBody = (0, converters_1.schema2value)(content.schema, true);
                    }
                }
                if (reqFormat) {
                    params.push({
                        name: 'reqFormat',
                        required: true,
                        description: null,
                        values: [
                            {
                                isArray: false,
                                isEnum: false,
                                nullable: false,
                                description: null,
                                value: reqFormat
                            }
                        ]
                    });
                }
                if (reqBody) {
                    params.push({
                        name: 'reqBody',
                        required: required,
                        description: description,
                        values: [reqBody]
                    });
                }
            }
            return {
                name: method,
                required: true,
                description: (_o = target.description) !== null && _o !== void 0 ? _o : null,
                values: [
                    { isArray: false, isEnum: false, nullable: false, description: null, value: params }
                ]
            };
        })
            .filter(function (method) { return !!method; });
        if (methods.length) {
            var methodsText = (0, props2String_1.props2String)(methods, '');
            var hasBinary = methodsText.includes(converters_1.BINARY_TYPE);
            var hasTypes = /( |<)Types\./.test(methodsText);
            return {
                file: file,
                methods: "/* eslint-disable */\n".concat(hasBinary ? "import type { ReadStream } from 'fs'\n" : '').concat(hasBinary && !hasTypes ? '\n' : '').concat(hasTypes
                    ? "import type * as Types from '".concat(file.map(function () { return ''; }).join('../'), "@types'\n\n")
                    : '', "export type Methods = ").concat(methodsText, "\n")
            };
        }
        else {
            return { file: file, methods: '' };
        }
    })
        .filter(function (file) { return file.methods; })), false));
    var typesText = parameters.length + schemas.length + requestBodies.length + responses.length
        ? __spreadArray(__spreadArray(__spreadArray(__spreadArray([], __read(parameters.map(function (p) { return ({
            name: p.name,
            description: null,
            text: typeof p.prop === 'string' ? p.prop : (0, props2String_1.props2String)([p.prop], '')
        }); })), false), __read(schemas.map(function (s) { return ({
            name: s.name,
            description: s.value.description,
            text: (0, props2String_1.value2String)(s.value, '').replace(/\n {2}/g, '\n')
        }); })), false), __read(requestBodies.map(function (r) { return ({
            name: r.name,
            description: null,
            text: typeof r.value === 'string'
                ? r.value
                : (0, props2String_1.value2String)(r.value, '').replace(/\n {2}/g, '\n')
        }); })), false), __read(responses.map(function (r) { return ({
            name: r.name,
            description: null,
            text: typeof r.value === 'string'
                ? r.value
                : (0, props2String_1.value2String)(r.value, '').replace(/\n {2}/g, '\n')
        }); })), false).map(function (p) { return "\n".concat((0, props2String_1.description2Doc)(p.description, ''), "export type ").concat(p.name, " = ").concat(p.text, "\n"); })
            .join('')
            .replace(/(\W)Types\./g, '$1')
            .replace(/\]\?:/g, ']:')
        : null;
    return {
        openapi: openapi,
        baseURL: ((_f = (_e = openapi.servers) === null || _e === void 0 ? void 0 : _e[0]) === null || _f === void 0 ? void 0 : _f.url) || '',
        types: typesText &&
            "/* eslint-disable */".concat(typesText.includes(converters_1.BINARY_TYPE) ? "\nimport type { ReadStream } from 'fs'\n" : '').concat(typesText),
        files: files
    };
});
